backend/metadata_writer/src/metadata_writer.py

```diff
@@ -0,0 +1,756 @@
+                    input_name = argo_artifact.get('path', '') # Every artifact should have a path in Argo
+                    input_artifact_path_prefix = '/tmp/inputs/'
+                    input_artifact_path_postfix = '/data'
+                    if input_name.startswith(input_artifact_path_prefix):
+                        input_name = input_name[len(input_artifact_path_prefix):]
+                    if input_name.endswith(input_artifact_path_postfix):
+                        input_name = input_name[0: -len(input_artifact_path_postfix)]
+
+                    artifact = link_execution_to_input_artifact(
+                        store=mlmd_store,
+                        execution_id=execution.id,
+                        uri=artifact_uri,
+                        input_name=input_name,
+                    )
+                    if artifact is None:
+                        # TODO: Maybe there is a better way to handle missing upstream artifacts
+                        continue
+
+                    input_artifact_ids.append(dict(
+                        id=artifact.id,
+                        name=input_name,
+                        uri=artifact.uri,
+                    ))
+                    print('Found Input Artifact: ' + str(dict(
+                        input_name=input_name,
+                        id=artifact.id,
+                        uri=artifact.uri,
+                    )))
+
+                execution_id = execution.id
+                context_id = run_context.id
+
+                obj.metadata.labels[METADATA_EXECUTION_ID_LABEL_KEY] = execution_id
+                obj.metadata.labels[METADATA_CONTEXT_ID_LABEL_KEY] = context_id
+
+                metadata_to_add = {
+                    'labels': {
+                        METADATA_EXECUTION_ID_LABEL_KEY: str(execution_id),
+                        METADATA_CONTEXT_ID_LABEL_KEY: str(context_id),
+                    },
+                    'annotations': {
+                        METADATA_INPUT_ARTIFACT_IDS_ANNOTATION_KEY: json.dumps(input_artifact_ids),
+                    },
+                }
+
+                patch_pod_metadata(
+                    namespace=obj.metadata.namespace,
+                    pod_name=obj.metadata.name,
+                    patch=metadata_to_add,
+                )
+                pod_name_to_execution_id[obj.metadata.name] = execution_id
+                workflow_name_to_context_id[argo_workflow_name] = context_id
+
+                print('New execution id: {}, context id: {} for pod {}.'.format(execution_id, context_id, obj.metadata.name))
+
+                print('Execution: ' + str(dict(
+                    context_id=context_id,
+                    context_name=argo_workflow_name,
+                    execution_id=execution_id,
+                    execution_name=obj.metadata.name,
+                    component_name=component_name,
+                )))
+
+                # TODO: Log input parameters as execution options.
```

Comment:
Maybe file an issue so that it gets triaged.

IMO this is not very hard, right?

Reply:
This is harder than it might seem (see the "Future work" section of the design doc).
To log execution properties, they must be registered in the ExecutionType property schema.
Unfortunately, ContainerOp does not have concept of inputs and the input names (or even their existence) are not preserved. So, there are no stable input names and even the number of "recoverable" inputs is not stable. So with the naive implementation, the ExecutionType property schema will be different for each task, but MLMD won't allow that.
To fix this issue, we first need to keep/recover/stabilize the input information when it passes through ContainerOp. This is not trivial, especially for the cases where the users directly construct ContainerOps and not use components.
Another solution would be to somehow start passing all data as artifacts, effectively eliminating the execution properties.

Reply:
Thanks for the explanation. I'm good with the way it is now.